tests/unit/test_characterProcessing.py

```diff
@@ -0,0 +1,74 @@
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2020 NV Access Limited
+
+"""Unit tests for the characterProcessing module.
+"""
+
+import unittest
+import re
+from characterProcessing import SYMLVL_ALL
+from characterProcessing import SpeechSymbolProcessor
+from characterProcessing import processSpeechSymbols as process
+
+
+class TestComplex(unittest.TestCase):
+	"""Test the complex symbols rules.
+	"""
+
+	def _replace_cb(self, name, replacement):
+		"""Return a regexp callback which replaces matches of the given
+		group name with the replacement string, with support for
+		replacement of group references.
+		"""
+		def replace(self, m):
+			if m.lastgroup == name:
+				return SpeechSymbolProcessor._replaceGroups(self, m, replacement)
+			return m.group()
+		return self.replace
+
+	def _replace(self, string, pattern, name, replacement):
+		"""Perform a pattern-based replacement on a string, for one
+		named group, with support for replacement of group references.
+		"""
+		regexp = re.compile(pattern, re.UNICODE)
+		return regexp.sub(self._replace_cb(name, replacement), string)
+
+	def test_group_replacement(self):
+		"""Test that plain text gets properly replaced
+		"""
+		self.assertEqual(self._replace("1", r"(?P<foo>(\d))", "foo", "a"), "a")

```

Comment:
Thanks for adding extra tests, at the moment it is hard to read the 'actual' and 'expected' values.There is a lot happening on these lines. Could you split them in the following way please?
```suggestion
		replaced = self._replace(
			string="1",
			pattern=r"(?P<foo>(\d))",  # name group 'foo' looking for a single digit.
			name="foo",
			replacement="a"
		)
		self.assertEqual(replaced, "a")

```

Reply:
Ok, done so